### PR TITLE
rustdoc: Stop textually replacing `Self` in doc links before resolving them

### PR DESCRIPTION
**Pull request by petrochenkov**
Resolve it directly to a type / def-id instead.

Also never pass `Self` to `Resolver`, it is useless because it's guaranteed that no resolution will be found.

This is a pre-requisite for https://github.com/rust-lang/rust/issues/83761.

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks! Better have @camelid take a look too.

**Comment by camelid** [comment]
Thanks for doing this! I've been wanting to fix this hack for a while. I'll review this soon.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -499,19 +502,87 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+                item_id,
                 module_id,
                 partial_res: None,
                 unresolved: path_str.into(),
             })
         })
     }
 
+    fn resolve_self_ty(&self, path_str: &str, ns: Namespace, item_id: ItemId) -> Option<Res> {
+        if ns != TypeNS || path_str != "Self" {
```
Comment (camelid) [inline src/librustdoc/passes/collect_intra_doc_links.rs:514]: Why have `ns` if it's always `TypeNS`? Likewise with `path_str`.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -499,19 +502,87 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+        }
+
+        let self_id = match item_id.as_def_id() {
+            None => None,
+            Some(did)
+                if (matches!(self.cx.tcx.def_kind(did), DefKind::Field)
+                    && matches!(
+                        self.cx.tcx.def_kind(self.cx.tcx.parent(did).unwrap()),
+                        DefKind::Variant
+                    )) =>
```
Comment (camelid) [inline src/librustdoc/passes/collect_intra_doc_links.rs:525]: Could you please make these regular if statements rather than guards?

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 3]
```diff
@@ -499,19 +502,87 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+                        Res::Def(kind, adt_def.did)
+                    }
+                    ty::Foreign(def_id) => Res::Def(DefKind::ForeignTy, *def_id),
+                    ty::Int(int_ty) => Res::Primitive((*int_ty).into()),
+                    ty::Uint(uint_ty) => Res::Primitive((*uint_ty).into()),
+                    ty::Float(float_ty) => Res::Primitive((*float_ty).into()),
+                    ty::Bool => Res::Primitive(PrimitiveType::Bool),
+                    ty::Char => Res::Primitive(PrimitiveType::Char),
+                    ty::Str => Res::Primitive(PrimitiveType::Str),
+                    _ => Res::Def(def_kind, self_id),
```
Comment (camelid) [inline src/librustdoc/passes/collect_intra_doc_links.rs:563]: What about all the other primitives, for example slices?

Comment (camelid) [reply src/librustdoc/passes/collect_intra_doc_links.rs:563]: This defkind is wrong as well.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 4]
```diff
@@ -723,10 +798,27 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+                                        root_res,
+                                        ItemFragment(FragmentKind::Variant, variant.def_id),
+                                    ));
+                                }
+                            }
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+
```
Comment (camelid) [inline src/librustdoc/passes/collect_intra_doc_links.rs:821]: Why this change?

File: src/test/rustdoc/intra-doc/associated-items.rs [hunk 5]
```diff
@@ -57,4 +57,12 @@ impl T2 for S {
     fn ambiguous_method() {}
 }
 
+// @has associated_items/enum.MyEnum.html '//a/@href' 'enum.MyEnum.html#variant.MyVariant'
+/// Link to [MyEnumAlias::MyVariant]
```
Comment (camelid) [inline src/test/rustdoc/intra-doc/associated-items.rs:61]: This link is a valid Rust path right? Also I assume this is something that didn't work in intra-doc links before?

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 4 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:821]: Because `EnumTypeAlias::Variant` is a legal path, and rustdoc didn't support it.
`Self::Variant` is a special case of the above, with `Self` being the type alias.

File: src/test/rustdoc/intra-doc/associated-items.rs (hunk 5 above)
Comment (petrochenkov) [reply src/test/rustdoc/intra-doc/associated-items.rs:61]: Yes, https://github.com/rust-lang/rust/pull/93805#discussion_r805119510.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:514]: Just to keep all the `Self` resolution logic in a single function.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 2 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:525]: Ok, I just copy-pasted the previously existing code here.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 3 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:563]: Majority of the code in this file is wrong to some extent, to be honest.
So my additions keep their hackiness roughly on the same level with the surrounding code, otherwise you'll have to rewrite everything here to do things correctly.

Here I only need to consider inherent impls that are later considered by `resolve_primitive_associated_item`.
And that's all the cases listed in rustdoc's `enum PrimitiveType` as I understand?
And more specifically, the variants used in `fn all_impls`.
In that case I need to extend this list, yes.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:514]: Also, `ns` shouldn't ideally be only `TypeNS`, `Self` in value namespace should also be resolved to constructors if we are doing things correctly.

But that also requires some more logic in ` fn resolve_with_disambiguator` (where we accumulate resolutions from all namespaces) to avoid reporting false ambiguities (similar to `is_derive_trait_collision`).
I'm not sure why other "type vs its constructors" collisions are not processed in the same place.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 2 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: Done.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 3 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: Done.
I've reused `fn def_id_to_res` here.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 6]
```diff
@@ -343,6 +346,7 @@ impl ItemFragment {
 
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
 struct ResolutionInfo {
+    item_id: ItemId,
     module_id: DefId,
```
Comment (camelid) [inline src/librustdoc/passes/collect_intra_doc_links.rs:350]: Just wondering, could we remove `module_id` now that we have `item_id`? Not necessary to address now, but I wanted to bring it up.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 7]
```diff
@@ -1242,19 +1260,8 @@ impl LinkCollector<'_, '_> {
...
         let is_lone_crate = path_str == "crate";
-        if path_str.starts_with("Self::") || is_lone_self {
-            if let Some(ref name) = self_name {
-                if is_lone_self {
-                    path_str = name;
-                } else {
-                    resolved_self = format!("{}::{}", name, &path_str[6..]);
-                    path_str = &resolved_self;
-                }
-            }
```
Comment (camelid) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1256]: Where are paths starting with `Self::` handled in the new code? I only see explicit handling of `Self` (i.e., "lone Self"), and I'm a bit unclear about the code path `Self::` follows.

**Review by camelid — COMMENTED** [review]
Sorry this took so long. I only have one question; other than that, this looks great!

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 6 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:350]: Not without replacing it with some extra data, I tried it but wasn't able to do it right away.

The first issue is that rustdoc (unlike rustc) resolves paths differently in inner and outer attributes.
So `item_id` will be the same for inner and outer paths, but their `module_id`s will be different.

There are also may be nuances with how `item_id` and `module_id` are assigned for reexports, but I didn't investigate it.

I agree that ideally we would have an `item_id` and some additional orthogonal flags here instead of the `module_id`.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 7 above)
Comment (petrochenkov) [reply src/librustdoc/passes/collect_intra_doc_links.rs:1256]: `Self::something` paths are always resolved as type-relative associated paths (with `Self` being the type), in the same way as `TypeAlias::something`.

Comment (camelid) [reply src/librustdoc/passes/collect_intra_doc_links.rs:1256]: Ah, I see.

**Review by camelid — APPROVED** [review]
Thanks again!

**Comment by camelid** [comment]
Wow, that regression is way higher than I expected. I thought this PR might _improve_ perf. I still think this change is worth the regression, but do you know what caused the regression? Maybe the extra `item_id` fields?

**Comment by petrochenkov** [comment]
I'll try to benchmark this, but not earlier than landing the last part of https://github.com/rust-lang/rust/issues/83761.

**Comment by pnkfelix** [comment]
It looks to me like collect_intra_doc_links got [8.8% slower](https://perf.rust-lang.org/detailed-query.html?sort_idx=-11&commit=1661e4c7e0e68b4297aec095064d80566d4ea2b1&base_commit=0cbef1c6a7f4ee33cd41d91778544c5399f10f39&benchmark=serde-doc&scenario=full) on serde doc. It took [0.137 s](https://perf.rust-lang.org/detailed-query.html?commit=0cbef1c6a7f4ee33cd41d91778544c5399f10f39&benchmark=serde-doc&scenario=full) before this PR, and [0.149 s](https://perf.rust-lang.org/detailed-query.html?commit=1661e4c7e0e68b4297aec095064d80566d4ea2b1&benchmark=serde-doc&scenario=full) after this PR.



If I had to guess, I would assume the slowdown is due to the extra `item_id` field that was added to `ResolutionInfo`; the latter is used as a [key in a hashtable](https://github.com/petrochenkov/rust/blob/25c5e39bb14976a969664d450086b6e8dc09e318/src/librustdoc/passes/collect_intra_doc_links.rs#L378) in the `LinkCollector`. (maybe the size of `ResolutionInfo` passed some threshold that causes cache line problems? Just guessing at this point.)

Its funny that the slowdown only registered on serde doc though.

If you want to look further, @petrochenkov , then that would be great. But I don't think we need to spend too much time dissecting this, and I'm going to treat it as triaged.

@rustbot label: +perf-regression-triaged

**Comment by camelid** [comment]
`ResolutionInfo` did increase from 64 bytes to 80 bytes, so the cache line theory sounds quite plausible (e.g., since it's no longer a power of two).
